tfjs-backend-webgpu/src/depthwise_conv2d_webgpu.ts

```diff
@@ -65,70 +66,74 @@ export class DepthwiseConv2DProgram implements WebGPUProgram {
-              }
-            }
-          } else {
-            for (var wR = 0; wR < uniforms.filterHeight; wR = wR + 1) {
-              let xR = inputRowStart + wR * uniforms.dilation[0];
-
-              if (xR < 0 || xR >= uniforms.inDims[0]) {
-                continue;
+          let d2 = coords[${this.isChannelsLast ? 3 : 1}];
+          let channelMul = uniforms.wShape[3];
+          let d1 = d2 / channelMul;
+          let q = d2 % channelMul;
+
+          let inputRowStart = xRCCorner.x;
+          let inputColStart = xRCCorner.y;
+          let inputRowEnd = inputRowStart + uniforms.filterHeight *
+              uniforms.dilation[0];
+          let inputColEnd = inputColStart + uniforms.filterWidth *
+              uniforms.dilation[1];
+
+          // Convolve x(?, ?, d1)|x(d1, ?, ?) with w(:, :, d1, q) to get
+          // y(yR, yC, d2)|y(d2, yR, yC). ? = to be determined. : = across all
+          // values in that axis. x(?, ?, d1) and y(yR, yC, d2) is for NHWC.
+          // x(d1, ?, ?) and y(d2, yR, yC) is for NCHW.
+          var value = 0.0;
+
+          // Extract if checking out of for loop for performance.
+          if (inputRowStart >= 0 && inputColStart >= 0 &&
+            inputRowEnd < uniforms.inDims[0] &&
+                inputColEnd < uniforms.inDims[1]) {
+              for (var wR = 0; wR < uniforms.filterHeight; wR = wR + 1) {
+                let xR = inputRowStart + wR * uniforms.dilation[0];
+
+                for (var wC = 0; wC < uniforms.filterWidth; wC = wC + 1) {
+                  let xC = inputColStart + wC * uniforms.dilation[1];
+
+                  let xVal = ${getXSnippet};
+                  let wVal = getW(wR, wC, d1, q);
+                  value = value + xVal * wVal;
+                }
               }
+            } else {
+              for (var wR = 0; wR < uniforms.filterHeight; wR = wR + 1) {
+                let xR = inputRowStart + wR * uniforms.dilation[0];
 
-              for (var wC = 0; wC < uniforms.filterWidth; wC = wC + 1) {
-                let xC = inputColStart + wC * uniforms.dilation[1];
-
-                if (xC < 0 || xC >= uniforms.inDims[1]) {
+                if (xR < 0 || xR >= uniforms.inDims[0]) {
                   continue;
                 }
 
-                let xVal = ${getXSnippet};
-                let wVal = getW(wR, wC, d1, q);
-                value = value + xVal * wVal;
+                for (var wC = 0; wC < uniforms.filterWidth; wC = wC + 1) {
+                  let xC = inputColStart + wC * uniforms.dilation[1];
+
+                  if (xC < 0 || xC >= uniforms.inDims[1]) {
+                    continue;
+                  }
+
+                  let xVal = ${getXSnippet};
+                  let wVal = getW(wR, wC, d1, q);
+                  value = value + xVal * wVal;
+                }
               }
             }
+            ${biasActivationSnippet(this.addBias, this.activation)}
+          if (coordsInBounds4D(coords, uniforms.outShape)) {
```

Comment:
Since you already used `if (index < uniforms.size)`, `if (coordsInBounds4D(coords, uniforms.outShape))` is not needed anymore.

Reply:
Thanks, updated.